pkg/controller/combo_controller.go

```diff
@@ -2,36 +2,183 @@ package controller
+	templateHandler := handler.EnqueueRequestsFromMapFunc(c.mapTemplateToCombinations)
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.Combination{}).
+		Watches(&source.Kind{Type: &v1alpha1.Template{}}, templateHandler).
 		Complete(c)
 }
 
+// mapTemplateToCombinations is responsible for taking the template object and finding all associated
+// combinations that should be requeued. This should only happen whenever a template is changed in someway.
+func (c *combinationController) mapTemplateToCombinations(template client.Object) []reconcile.Request {
+	if template == nil {
+		return nil
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	requests := []reconcile.Request{}
+
+	// Build the label selector for querying
+	templateName := template.GetName()
+	templateSelector, err := labels.Parse("ReferencedTemplate=" + templateName)
+	if err != nil {
+		return requests
+	}
+
+	// Find all of the combinations that rely on this template and enqueue them for updates
+	combinationList := v1alpha1.CombinationList{}
+	if err := c.List(ctx, &combinationList, &client.ListOptions{LabelSelector: templateSelector}); err != nil {
+		return requests
+	}
+
+	for _, combination := range combinationList.Items {
+		c.log.Info(fmt.Sprintf("enqueueing %s combination in response to associated %s template being updated", combination.Name, templateName))
+		requests = append(requests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      combination.Name,
+				Namespace: template.GetNamespace(),
+			},
+		})
+	}
+
+	return requests
+}
+
+// Reconcile manages incoming combination CR's and processes them accordingly
 func (c *combinationController) Reconcile(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
-	// Set up a convenient log object so we don't have to type request over and over again
+	// Set up a convenient log object so we don’t have to type request over and over again
 	log := c.log.WithValues("request", req)
-	log.Info("reconciling combination")
 
-	in := &v1alpha1.Combination{}
-	if err := c.Get(ctx, req.NamespacedName, in); err != nil {
+	log.V(1).Info("new combination event inbound")
+
+	// Attempt to retrieve the requested combination CR
+	combination := &v1alpha1.Combination{}
+	err := c.Get(ctx, req.NamespacedName, combination)
+	if err != nil {
 		return reconcile.Result{}, client.IgnoreNotFound(err)
 	}
 
-	return reconcile.Result{}, nil
+	// If combination is being deleted, remove from queue
+	if !combination.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("combination is being deleted, ignoring event")
+		return reconcile.Result{}, nil
+	}
+
+	// Remove any previous evaluation in case of failure
+	combination.Status.Evaluations = []string{}
+
+	// Apply label for the referenced template for easy watching
+	if combination.Labels == nil {
+		combination.Labels = map[string]string{}
+	}
+	combination.Labels["ReferencedTemplate"] = combination.Spec.Template
```

Comment:
2 notes:
1. Template names can exceed the maximum length of label values (63 characters). This can be captured in an issue and fixed before a minor release.
2. Typically the label's key should include information about the controller that added it, see this [example](https://github.com/operator-framework/operator-lifecycle-manager/blob/c6cc1d7f706d355e34b1fc6523693bc13ceb4650/pkg/controller/operators/catalog/subscriptions.go#L13-L18) of multiple labels used by OLM.

Reply:
[Here](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) is some additional reading on label requirements in case you want to learn more.

Reply:
Most recent push changed this code to contain the following.

For # 1:
```
// Retrieve and validate the template's name
templateName := template.GetName()
if isValid := validation.IsValidLabelValue(templateName); isValid != nil {
	c.log.Info(fmt.Sprintf("template referenced in mapTemplateToCombinations call is not valid: %v", isValid))
	return requests
}
```

For # 2
```
// Build the label selector for querying
templateSelector, err := labels.Parse("combo.ReferencedTemplate=" + templateName)
if err != nil {
	return requests
}
```
This will address the issues you outlined in terms of validation and best practice. Thoughts?